### PR TITLE
Add Lifetime to vir TypeVar and Reference

### PR DESCRIPTION
**Pull request by pascal-huber**
Add Lifetime to vir TypeVar and Reference

# Work to do

 - [x] Add types to vir high
 - [x] Fix issues in vir_gen

### CODE REVIEW

**Comment by vakaras** [comment]
`vir/defs/high/operations_internal/identifier/ty.rs` needs:

```rust
impl WithIdentifier for ty::TypeVar {
    fn get_identifier(&self) -> String {
        match self {
            ty::TypeVar::Lifetime(type_var) => type_var.get_identifier(),
            ty::TypeVar::GenericType(type_var) => type_var.get_identifier(),
        }
    }
}

impl WithIdentifier for ty::Lifetime {
    fn get_identifier(&self) -> String {
        self.name.clone()
    }
}

impl WithIdentifier for ty::GenericType {
    fn get_identifier(&self) -> String {
        self.name.clone()
    }
}
```

File: prusti-viper/src/encoder/mir/generics/interface.rs [hunk 1]
```diff
@@ -113,6 +113,6 @@ impl<'v, 'tcx: 'v> MirGenericsEncoderInterface<'tcx> for super::super::super::En
     }
     fn encode_param(&self, name: Symbol, index: u32) -> vir_high::ty::TypeVar {
         let identifier = format!("{}${}", name.as_str(), index);
-        vir_high::ty::TypeVar::new(identifier)
+        vir_high::ty::TypeVar::GenericType(vir_high::ty::GenericType { name: identifier })
```
Comment (vakaras) [inline prusti-viper/src/encoder/mir/generics/interface.rs]:
```suggestion
        vir_high::ty::TypeVar::generic_type(identifier)
```

**Review by vakaras — APPROVED** [review]
LGTM.

bors r+

**Comment by vakaras** [comment]
@pascal-huber To format code, run `./x.py fmt-all`.

**Comment by pascal-huber** [comment]
> @pascal-huber To format code, run `./x.py fmt-all`.

done

File: prusti-viper/src/encoder/mir/generics/interface.rs (hunk 1 above)
Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/generics/interface.rs]: That doesn't seem to work/compile. Should `generic_type` be generated?

Comment (vakaras) [reply prusti-viper/src/encoder/mir/generics/interface.rs]: Try adding `#[derive_helpers]` on the `TypeVar` definition.

Comment (pascal-huber) [reply prusti-viper/src/encoder/mir/generics/interface.rs]: thanks, done

**Pull request merged**